### PR TITLE
[shortfin] Add argmax host op.

### PR DESCRIPTION
**Pull request by stellaraccident**
Adds `shortfin.array.argmax`, following closely to numpy conventions.

In the C++ implementation, this forced completion of some APIs around accessing shortfin arrays via xtensor. 

C++ based users will just use the `device_array::map_xtensor*` helpers to access xtensor expressions directly.

For Python, we export select eager op kernels to match common situations. The intent here is to just be able to do some lightweight algebra and data manipulation common in host-side ML (without deps).

### CODE REVIEW

File: libshortfin/src/shortfin/array/dims.h [hunk 1]
```diff
@@ -95,6 +137,10 @@ class SHORTFIN_API InlinedDims {
       std::fill(dims_.inline_dims.begin(), dims_.inline_dims.end(), value);
     }
   }
+  template <std::contiguous_iterator BeginTy, typename EndTy>
+  InlinedDims(BeginTy begin, EndTy end) {
```
Comment (Groverkss) [inline libshortfin/src/shortfin/array/dims.h:141]: nit: assert end > begin

File: libshortfin/tests/api/array_ops.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    assert out.view(3).items.tolist() == [0] * 16
+
+    # out= keepdims variant
+    out = sfnp.device_array(device, [4, 16, 1], dtype=sfnp.int64)
+    sfnp.argmax(src, keepdims=True, out=out)
+    assert out.shape == [4, 16, 1]
+    assert out.view(0).items.tolist() == [127] * 16
+    assert out.view(1).items.tolist() == [0] * 16
+    assert out.view(2).items.tolist() == [127] * 16
+    assert out.view(3).items.tolist() == [0] * 16
```
Comment (Groverkss) [inline libshortfin/tests/api/array_ops.py:68]: Can we have an example of argmax with axis=0?

File: libshortfin/python/array_host_ops.cc [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace {
+
+static const char DOCSTRING_ARGMAX[] =
+    R"(Returns the indices of the maximum values along an axis.
+
+See documentation for numpy.argmax which this operator emulates.
+
+Implemented dtypes: float16, float32.
+)";
```
Comment (Groverkss) [inline libshortfin/python/array_host_ops.cc:26]: Can we add a note somewhere that out must be / will be int64? It's not obvious here. Also, would be good to document what divise_visible means.

Also, maybe differing opinion, personally, I prefer the docs to not delegate semantics to something else. I don't think argmax docs are going to change much, can we just copy them here from numpy and mention the semantics mimic numpy?

Also, does nanobind allow specifying parameter specific docs? I like that numpy docs has a parameter section for each operator and has per parameter docs.

**Review by Groverkss — APPROVED** [review]

**Comment by Groverkss** [comment]
> On the C++ side, this forced completion of some APIs around accessing shortfin arrays via xtensor. C++ based code will just use xtensor directly vs having a built-in op library.

> For Python, we export select eager op kernels to match common situations. The intent here is to just be able to do some lightweight algebra and data manipulation common in host-side ML (without deps).

What does the c++ / python side mean? Does it mean libshortfin c++ side and libshortfin python side, or do you mean host argmax vs device argmax? I'm a bit confused because I just see argmax bindings simply calling the c++ implementation.

**Comment by stellaraccident** [comment]
> > On the C++ side, this forced completion of some APIs around accessing shortfin arrays via xtensor. C++ based code will just use xtensor directly vs having a built-in op library.
> 
> > For Python, we export select eager op kernels to match common situations. The intent here is to just be able to do some lightweight algebra and data manipulation common in host-side ML (without deps).
> 
> What does the c++ / python side mean? Does it mean libshortfin c++ side and libshortfin python side, or do you mean host argmax vs device argmax? I'm a bit confused because I just see argmax bindings simply calling the c++ implementation.

I'll make it clearer. What I meant was that if coding against shortfin and needing to do host side array manipulation in C++, users can just bail to xtensor directly and don't need type erased op-by-op wrappers like we provide for Python. If you're in C++, that will be superior because the xtensor expression trees use compile time code generation that you lose if you erase element types.

If coding as a python user and needing to do array manipulation, we wrap up some set of these ops so basic things can be done without more deps.

**Pull request merged**